flink-connector-pinot/src/test/java/org/apache/flink/streaming/connectors/pinot/PinotSinkTest.java

```diff
@@ -0,0 +1,396 @@
+        DataStream<SingleColumnTableRow> dataStream = setupStreamingDataSource(env, rawData);
+        // With a segment size of MAX_ROWS_PER_SEGMENT = 5 elements and a parallelism of 2,
+        // the failure will be raised once the first 2 segments were committed to Pinot.
+        dataStream = setupFailingMapper(dataStream, 12);
+        setupSink(dataStream);
+
+        // Start execution of job
+        env.executeAsync();
+
+        // Check for data in Pinot
+        checkForDataInPinotWithRetry(rawData);
+
+        // Generator source can now shut down
+        latch.countDown();
+    }
+
+    /**
+     * Generates a small test dataset consisting of {@link SingleColumnTableRow}s.
+     *
+     * @return List of SingleColumnTableRow
+     */
+    private List<String> getRawTestData(int numItems) {
+        return IntStream.range(1, numItems + 1)
+                .mapToObj(num -> "ColValue" + num)
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * Setup the data source for STREAMING tests.
+     *
+     * @param env           Stream execution environment
+     * @param rawDataValues Data values to send
+     * @return resulting data stream
+     */
+    private DataStream<SingleColumnTableRow> setupStreamingDataSource(StreamExecutionEnvironment env, List<String> rawDataValues) {
+        SimpleStreamingSource source = new SimpleStreamingSource(rawDataValues, 10);
+        return env.addSource(source)
+                .name("Test input");
+    }
+
+    /**
+     * Setup the data source for BATCH tests.
+     *
+     * @param env           Stream execution environment
+     * @param rawDataValues Data values to send
+     * @return resulting data stream
+     */
+    private DataStream<SingleColumnTableRow> setupBatchDataSource(StreamExecutionEnvironment env, List<String> rawDataValues) {
+        return env.fromCollection(rawDataValues)
+                .map(value -> new SingleColumnTableRow(value, System.currentTimeMillis()))
+                .name("Test input");
+    }
+
+    /**
+     * Setup a mapper that fails when processing the nth element with n = failOnceAtNthElement.
+     *
+     * @param dataStream           Input data stream
+     * @param failOnceAtNthElement Number of elements to process before raising the exception
+     * @return resulting data stream
+     */
+    private DataStream<SingleColumnTableRow> setupFailingMapper(DataStream<SingleColumnTableRow> dataStream, int failOnceAtNthElement) {
+        AtomicInteger messageCounter = new AtomicInteger(0);
+
+        return dataStream.map(element -> {
+            if (!hasFailedOnce.get() && messageCounter.incrementAndGet() == failOnceAtNthElement) {
+                hasFailedOnce.set(true);
+                // Wait more than STREAMING_CHECKPOINTING_INTERVAL to ensure
+                // that at least one checkpoint was created before raising the exception.
+                Thread.sleep(4 * STREAMING_CHECKPOINTING_INTERVAL);
```

Comment:
Maybe you can find a way to also remove this timeout :) One idea would be to make your map function implement `CheckpointedFunction` which allows you to receive the checkpoint notification.

I also do not really know why it is important for the failure to have at-least one checkpoint.

Reply:
I've removed the timeout by extending the source to support exception raising and added the following comment to explain why we wait for a snapshot creation. I hope this clears things up.
> This allows to check whether the snapshot creation and failure recovery in
> {@link org.apache.flink.streaming.connectors.pinot.writer.PinotSinkWriter} works properly,
> respecting the already committed elements and those that are stored in an active
> {@link org.apache.flink.streaming.connectors.pinot.writer.PinotWriterSegment}. Committed 
> elements must not be saved to the snapshot while those in an active segment must be saved 
> to the snapshot in order to enable later-on recovery.